src/core/include/units/magnitude.h

```diff
@@ -480,6 +480,52 @@ constexpr auto operator*(magnitude<H1, T1...>, magnitude<H2, T2...>)
 
 constexpr auto operator/(Magnitude auto l, Magnitude auto r) { return l * pow<-1>(r); }
 
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Magnitude numerator and denominator implementation.
+
+namespace detail {
+
+// The largest integer which can be extracted from any magnitude with only a single basis vector.
+template<BasePower auto BP>
+constexpr auto integer_part(magnitude<BP>)
+{
+  constexpr auto power_num = numerator(BP.power);
+  constexpr auto power_den = denominator(BP.power);
+
+  if constexpr (std::is_integral_v<decltype(BP.get_base())> && (power_num >= power_den)) {
+    constexpr auto largest_integer_power = [power_num, power_den](BasePower auto bp) {
+      bp.power = (power_num / power_den);  // Note: integer division intended.
+      return bp;
+    }(BP);  // Note: lambda is immediately invoked.
+
+    return magnitude<largest_integer_power>{};
+  } else {
+    return magnitude<>{};
+  }
+}
+
+}  // namespace detail
+
+template<BasePower auto... BPs>
+constexpr auto numerator(magnitude<BPs...>)
+{
+  return (detail::integer_part(magnitude<BPs>{}) * ... * magnitude<>{});
+}
+
+constexpr auto denominator(Magnitude auto m) { return numerator(pow<-1>(m)); }
+
+// Implementation of conversion to ratio goes here, because it needs `numerator()` and `denominator()`.
+constexpr ratio as_ratio(Magnitude auto m)
+{
+  static_assert(is_rational(m));
```

Comment:
```suggestion
constexpr ratio as_ratio(Magnitude auto m)
  requires(is_rational(m))
{
```

Reply:
I love replacing the `static_assert` with a constraint, but g++ doesn't like this as written:

```
In file included from /home/chogg/github/units/test/unit_test/runtime/magnitude_test.cpp:24:
/home/chogg/github/units/src/core/include/units/magnitude.h:519:25: error: use of parameter outside function body before ‘)’ token
  519 |   requires(is_rational(m))
      |                         ^
/home/chogg/github/units/src/core/include/units/magnitude.h:519:12: error: there are no arguments to ‘is_rational’ that depend on a template parameter, so a declaration of ‘is_rational’ must be available [-fpermissive]
  519 |   requires(is_rational(m))
      |            ^~~~~~~~~~~
/home/chogg/github/units/src/core/include/units/magnitude.h:519:12: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
```

I can get it to work with `requires(is_rational(decltype(m){}))`.  We'll see how the other compilers like it!

Reply:
Which versions of GCC?

Reply:
GCC 10, I believe.